### ISSUE TITLE
RunDB / Optional uHash tracking issue

### ISSUE THREAD

**Issue by robled** [CentOS-PaaS-SIG/linchpin#430]
```
$ linchpin --version
linchpin version 1.2.1
```

During integration tests with cinch, the rundb feature is incompatible with the existing downstream example topologies that have been created for linchpin 1.0.4.  This command  - ```linchpin -v --creds-path ./credentials up```, when run in conjunction with the downstream example topologies (these will be provided offline), creates the following directory path which is not self-contained and cannot be run safely in Jenkins CI:

```
~/.config/linchpin 
```

Moving on from here, if we proceed with provisioning, our instance name (jmaster) as defined in our linchpin topology file has a UID appended to the end, once created:

```
jmaster-2f0c-1
```

At this point, if linchpin is run again, rather than idempotently checking that the initial instance was created, a new instance is created with a different UID:

```
jmaster-724f-1
```

Here, we would expect that the same instance  would be merely checked by linchpin/ansible for proper parameters, rather than creating a second instance.  

Moving forward again, the **inventories** directory of the linchpin workspace now contains inventory files that cinch cannot predict the names of:

```
$ ls -1 inventories/
cinch-test-2f0c.inventory
cinch-test-724f.inventory
cinch-test-9dee.inventory
cinch-test.inventory
```

The file named **cinch-test.inventory** (created by linchpin version 1.0.4) can be consumed by cinch for instance configuration, but the files with the UID appended to them cannot.

I suspect that these issues are related to the rundb feature, which can be disabled via **linchpin.conf**, however, we've already packaged a distribution of linchpin example workspaces with an existing **linchpin.conf**, and the impact of distributing that change at a large scale is something that we would like to avoid.

**Comment by herlo** [comment]
> ```
> $ linchpin --version
> linchpin version 1.2.1
> ```
>
> During integration tests with cinch, the rundb feature is incompatible with the existing downstream 
> example topologies that have been created for linchpin 1.0.4.  This command  
> - ```linchpin -v --creds-path ./credentials up```, when run in conjunction with the downstream example 
> topologies (these will be > provided offline), creates the following directory path which is not 
> self-contained and cannot be run safely in Jenkins CI:
>
> ```
> ~/.config/linchpin 
> ```

This value is however set here: https://github.com/CentOS-PaaS-SIG/linchpin/blob/develop/linchpin/linchpin.conf#L21, and could be updated to use the workspace, `/tmp` or another location quite easily. Something like:

`rundb_conn = {{ workspace }}/rundb.json`

In talking with others, it became clear that issue is not about the RunDB itself, but more about consistency across workspaces. Since it's possible that multiple linchpin actions could be happening at the same time, storing a single RunDB file in a central place may cause conflicts assuming the target names were the same in each instance.

I think adjusting the file path to the workspace is acceptable. Specifically, I'll file an issue to update this to write to the workspace in which linchpin is being run. It'll probably live in `{{ workspace }}/.rundb/rundb.json`. I'll make sure this directory is created automatically. The only constraint here is that if a user wishes for data records across workspaces, they would need to update to a more general location. 

> Moving on from here, if we proceed with provisioning, our instance name (jmaster) as defined in our
> linchpin topology file has a UID appended to the end, once created:
>
> ```
> jmaster-2f0c-1
> ```

> At this point, if linchpin is run again, rather than idempotently checking that the initial instance was 
> created, a new instance is created with a different UID:
>
>```
> jmaster-724f-1
> ```
>
> Here, we would expect that the same instance  would be merely checked by linchpin/ansible for proper 
> parameters, rather than creating a second instance.  

A request for unique resource names is responsible for this, see https://github.com/CentOS-PaaS-SIG/linchpin/issues/284 for more information. It was completed as part of the RunDB work.

What you didn't show is that an output is returned after the run. It contains the target, run_id, uHash, and Exit Code by default. Because being backward compatible was considered here, one can use this same resource over and over by simply obtaining the run_id and passing it to the linchpin command. Assuming the first run was had a run_id value of `1`, the following would provide idempotent (or repeatable) tasks.

`linchpin -v --creds-path ./credentials up -r 1`

This would also work for the `destroy` action.

The `linchpin journal` command was also written to gather this information as needed. 

What needs to happen here is a flag needs to be set to enable the uHash. It is currently the default, but that is not backward compatible. I'll file a separate issue to get this in place in the near future.

> Moving forward again, the **inventories** directory of the linchpin workspace now contains inventory 
> files that cinch cannot predict the names of:
>
> ```
> $ ls -1 inventories/
> cinch-test-2f0c.inventory
> cinch-test-724f.inventory
> cinch-test-9dee.inventory
> cinch-test.inventory
> ```

> The file named **cinch-test.inventory** (created by linchpin version 1.0.4) can be consumed by cinch 
> for instance configuration, but the files with the UID appended to them cannot.

> I suspect that these issues are related to the rundb feature, which can be disabled via **linchpin.conf**, 
> however, we've already packaged a distribution of linchpin example workspaces with an existing 
> **linchpin.conf**, and the impact of distributing that change at a large scale is something that we would 
> like to avoid.

One discussion recently had was to allow the `layout` to define the name of the inventory file. See https://github.com/CentOS-PaaS-SIG/linchpin/issues/421. This is still in discussion and design, but would provide a convenient way to get a consistent inventory name.

RunDB is just a datastore where the uHash (or UID as you describe above) is stored. It appears that if I were to update the uHash feature to optional as described above, the inventories would also be named as you expect, since the uHash would not be used. The uHash would continue to show up in the database and reporting, but could be safely ignored with these changes. 

Others require the unique values for their work, along with the `linchpin journal` functionality described above.

One other alternative to all of this is to create a linchpin hook for cinch. By default, there's a context provided to ansible hooks in linchpin which would automatically provide the inventory to the playbook. I know we've discussed this in the past, but maybe it's a good time to revisit the topic?

**Comment by herlo** [comment]
tagging @robled as requested.

**Comment by robled** [comment]
> What you didn't show is that an output is returned after the run. It contains the target, run_id, uHash, and Exit Code by default. Because being backward compatible was considered here, one can use this same resource over and over by simply obtaining the run_id and passing it to the linchpin command. Assuming the first run was had a run_id value of 1, the following would provide idempotent (or repeatable) tasks. 

>linchpin -v --creds-path ./credentials up -r 1

Changing the way the CLI is called would not be backwards-compatible.  We need to come up with a solution that would allow this [existing JJB template](https://github.com/RedHatQE/cinch/blob/master/jjb/ci-jslave-project-sample.yaml#L17) to work without changing the template as written for linchpin 1.0.4 and still be idempotent.

> One discussion recently had was to allow the layout to define the name of the inventory file. See #421. This is still in discussion and design, but would provide a convenient way to get a consistent inventory name.

It sounds like this would require a change to our existing layout files, which would not be backwards-compatible.

> One other alternative to all of this is to create a linchpin hook for cinch. By default, there's a context provided to ansible hooks in linchpin which would automatically provide the inventory to the playbook. I know we've discussed this in the past, but maybe it's a good time to revisit the topic?

cinch can't use hooks in linchpin because linchpin uses the Ansible API which, according the [Ansible docs](https://docs.ansible.com/ansible/latest/dev_guide/developing_api.html#python-api), is not stable.  This imposes limitations on which version of Ansible can be used in the virtualenv which contains linchpin.  cinch generally [tracks](https://github.com/RedHatQE/cinch/blob/master/setup.py#L32) the latest released version of Ansible, and [work has been completed](https://github.com/RedHatQE/cinch/pull/192) to give cinch its own virtualenv (separate from linchpin) with the latest released version of Ansible, so that both projects can be used together without running into dependency issues.

**Comment by herlo** [comment]
On Thu, Dec 14, 2017 at 1:26 PM, robled <notifications@github.com> wrote:

> What you didn't show is that an output is returned after the run. It
> contains the target, run_id, uHash, and Exit Code by default. Because being
> backward compatible was considered here, one can use this same resource
> over and over by simply obtaining the run_id and passing it to the linchpin
> command. Assuming the first run was had a run_id value of 1, the following
> would provide idempotent (or repeatable) tasks.
>
> linchpin -v --creds-path ./credentials up -r 1
>
> Changing the way the CLI is called would not be backwards-compatible. We
> need to come up with a solution that would allow this existing JJB
> template
> <https://github.com/RedHatQE/cinch/blob/master/jjb/ci-jslave-project-sample.yaml#L17>
> to work without changing the template as written for linchpin 1.0.4 and
> still be idempotent.
>
As I said above:

'What needs to happen here is a flag needs to be set to enable the uHash.
It is currently the default, but that is not backward compatible. I'll file
a separate issue to get this in place in the near future.'

Setting a flag called use_uHash = False in the linchpin.conf will
accomplish this once the work has been done. This will be the default, and
no changes to the linchpin.conf will be necessary from your end.

> One discussion recently had was to allow the layout to define the name of
> the inventory file. See #421
> <https://github.com/CentOS-PaaS-SIG/linchpin/issues/421>. This is still
> in discussion and design, but would provide a convenient way to get a
> consistent inventory name.
>
> It sounds like this would require a change to our existing layout files,
> which would not be backwards-compatible.
>

It would require a change to the layout files. It's just one option. As I
stated above, I also suggested that making the uHash optional, the
inventory file would be generated as expected.

... 'It appears that if I were to update the uHash feature to optional as
described above, the inventories would also be named as you expect, since
the uHash would not be used. The uHash would continue to show up in the
database and reporting, but could be safely ignored with these changes.'


> One other alternative to all of this is to create a linchpin hook for
> cinch. By default, there's a context provided to ansible hooks in linchpin
> which would automatically provide the inventory to the playbook. I know
> we've discussed this in the past, but maybe it's a good time to revisit the
> topic?
>
> cinch can't use hooks in linchpin because linchpin uses the Ansible API
> which, according the Ansible docs
> <https://docs.ansible.com/ansible/latest/dev_guide/developing_api.html#python-api>,
> is not stable. This imposes limitations on which version of Ansible can be
> used in the virtualenv which contains linchpin. cinch generally tracks
> <https://github.com/RedHatQE/cinch/blob/master/setup.py#L32> the latest
> released version of Ansible, and work has been completed
> <https://github.com/RedHatQE/cinch/pull/192> to give cinch its own
> virtualenv (separate from linchpin) with the latest released version of
> Ansible, so that both projects can be used together without running into
> dependency issues.
>
I have no problem with this, we can accomplish the same goals using
solutions described above.

**Comment by herlo** [comment]
@robled tagging here again

**Comment by robled** [comment]
Sounds good - based on my understanding (sans implementation details) it looks like at some point in the future we can run another integration test with the following expectations:

* ~`/.config/linchpin` will not be created
* existing invocations of the linchpin CLI will be unaffected, and existing behaviors, including idempotency, will be preserved
* the existing linchpin workspaces we've created for linchpin 1.0.4 can continue to run without being modified
* instance names and inventory output files will remain the same as they were in linchpin 1.0.4

**Comment by arilivigni** [comment]
@herlo Since there was agreement 27 days ago is there a fix for this in 1.5rc or 1.5.  This is currently blocking the compatibility with cinch and preventing them from moving to this version.  The main objective was not to have a 1.2.X maintenance branch but to allow backward compatibility in 1.5 for cinch and other that depend on the layouts, topology, etc.

What is the current ETA of 1.5 and this issue being addressed in it?

**Comment by herlo** [comment]
Both issues #433 and #434 have now been fixed with this issue. When v1.5.1 is released, this issue will be closed.

**Comment by herlo** [comment]
Fixed by PR #465 #462 #458
